### ISSUE TITLE
problems with thermo elastic drift from file

### ISSUE THREAD

**Issue by BertoltLang** [IvS-KULeuven/PlatoSim3#293]
I wasn't able to perform a simulation with drift from file activated. I ran into the

WARNING Telescope: updateTelescopeOrientation() at time before previous request: Not Implemented. 

message over and over again.

I could solve the problem for me, when I altered the function ThermoElasticDriftFromFile::ThermoElasticDriftFromFile(ConfigurationParameters &configParams) to work like the JitterFromFile::JitterFromFile(ConfigurationParameters &configParams) function.

Thanks in advance,

Bert

**Comment by JorisDeRidder** [comment]
Could you please send us somehow (e.g. through WeTransfer) the drift file, so that we can understand where exactly the problem is located?

**Comment by BertoltLang** [comment]
I just used the airbus jitter file for test purposes, since jitter and drift share the same input modalities.

**Comment by JorisDeRidder** [comment]
Could you tell us which version of PlatoSim you are using, and which branch (master/develop)? We cannot reproduce this problem with the latest version of PlatoSim.

**Comment by BertoltLang** [comment]
I just tested this again with a new installation of PlatoSim fresh from the development branch and got the same problems as mentioned in this issue again.

I think the problem is the timeFromFile vector in ThermoElasticDriftFromFile.cpp. In my tests, the first time something is written to it is 'previousTime' (which was never initialized at that time but was zero in my tests) in line 64. The next entry is the first line from the drift file which is again zero.

So the two first entries for the timeFromFile vector in ThermoElasticDriftFromFile.cpp are both 0.0000. These values are doubles and thus afaik not exact. This leads for me to a negative heartbeatinterval of -0.00000 (which is defined as timeFromFile[1] - timeFromFile[0]), which means a negative timeStep in exposeDetector() in Camera.cpp and to the error.

I hope this was understandable, please correct me, if I got an error in my line of thinking.

**Comment by rhuygen** [comment]
I have just run a similar simulation where I use the Airbus Jitter file as DriftFromFile. All other parameters are left at default values. The simulation runs through without problems.

Can you provide your input file and other possible changes. Then I can try to reproduce your problem.

**Comment by BertoltLang** [comment]
[driftTest_0006.log](https://github.com/IvS-KULeuven/PlatoSim3/files/2556904/driftTest_0006.log)

[inputfgs.txt](https://github.com/IvS-KULeuven/PlatoSim3/files/2556906/inputfgs.txt)


the used drift file "shortJitter.txt" is just a cropped version of the airbus jitter file (only 3000 entries starting with time = 0.0)

**Comment by SaraRegibo** [comment]
Hi @BertoltLang I ran PlatoSim with your input file with only these differences:

* star catalogue → default value
* (RA, Dec) of pointing → default value
* drift file → Airbus jitter file
* dimensions of the pre-scan and over-scan should not be 0 (I get an error if they are) → I don't understand how this works for you

After building and running PlatoSim with these parameters, I still cannot reproduce the problem reported here.  A few remarks, though:

* The first couple of lines of the Airbus jitter file look like this for me (so no `t = 0` here):

```
   1.2500000e-01  -4.3900385e-02  -9.8717390e-02  -9.2585591e-02
   2.5000000e-01  -5.1318635e-02  -1.0298143e-01  -9.7153134e-02
   3.7500000e-01  -6.1854463e-02  -1.0302147e-01  -1.0313945e-01
   5.0000000e-01  -7.2505580e-02  -1.0254280e-01  -1.0754734e-01
   6.2500000e-01  -8.1132735e-02  -9.7395615e-02  -1.1030364e-01
```
Can you check whether these line also occur somewhere at the start of your file (apart from the lines with `t = 0`)?

* Another remark, unrelated to this issue, but good to know anyway.  You use a custom CCD position with `row = 0` as first exposed row.  Note that if you want frame transfer as for the Plato fast cameras, you would need `row = 2255` as your first exposed row.  This can be achieved by either a `Custom` CCD position and using 2255 as value for `FirstExposedRow`, or by choosing a pre-defined CCD position (1, 2, 3, 4).

**Comment by BertoltLang** [comment]
Hi Sara,

- I should have mentioned, that the inputfile I sent is the template that I use to start a simulation series with the demo_fgs.py script. The CCD position of the fast cameras is set this way.

- Comparing the values, we seem to use different jitter files. But I think the issue here is, that my files have an entry for time = 0.0. This should be the reason, why the program behaves the way I tried to describe yesterday.

- All files including time dependent values I use start with entries for time = 0.0. It would be great if PlatoSim could cope with this (like it already does in jitterFromFile.cpp)

**Comment by SaraRegibo** [comment]
Fixed by #321 (currently under review).

@JorisDeRidder Since you solved a similar problem with the jitter from file (negative heartbeat in some cases; see #246): could you have a look and review the code?  I am now using the exact same code now for the drift from file as for the jitter from file.  I don't see a reason why jitter and drift should be handled differently (but I could be wrong, of course)...

**Comment by SaraRegibo** [comment]
Implemented in `develop` by pull request #321.  Assigning to @BertoltLang for closure.

**Comment by BertoltLang** [comment]
Looks fine to me, thank you!
